### ISSUE TITLE
Installer should escape passwords

### ISSUE THREAD

**Issue by m4z** [SimpleMachines/SMF#5584]
#### Description
When running the installer, entering a database password containing `'` breaks the install process with various syntax errors, depending on the actual password. For example, using a 64-character KeePassXC-generated password starting with `$t&jz_vkCS$H*(32YsN~2D4'>JZur}$uAD$UjhX` resulted in:
```
==> /var/log/lighttpd/error.log <==
2019-04-07 19:33:43: (mod_fastcgi.c.2702) FastCGI-stderr: PHP Parse error:  syntax error, unexpected '}' in /home/some-username/public_html/Settings.php on line 104
```

Choosing the 1-character password `'` results in:
```
==> /var/log/lighttpd/error.log <==
2019-04-07 19:56:07: (mod_fastcgi.c.2702) FastCGI-stderr: PHP Parse error:  syntax error, unexpected '';' (T_CONSTANT_ENCAPSED_STRING) in /home/some-username/public_html/Settings.php on line 110
```

### Steps to reproduce
1. During installation, enter any database passwort containing `'`
2. Enjoy a beautiful "HTTP ERROR 500"

### Environment (complete as necessary)
- Version/Git revision: 2.1rc2 (on Debian 8.11 with lighttpd-1.4.35)
- Database Type: postgresql-9.4, but probably affects all
- Database Version: 9.4, but probably affects all
- PHP Version: 5.4, but probably affects all

### Additional information/references
Probably related to #2896.

Thanks for this otherwise (probably) great software! (Haven't yet managed to finish my first install. :wink: )

**Comment by m4z** [comment]
This seems to be more complex: Even if avoiding `'`, other special characters don't seem to be properly escaped, because I could not successfully connect with various passwords containing special characters.
I have chosen a password without special characters for now (which worked fine), but will post more details in a few days.

**Comment by live627** [comment]
I added a commit https://github.com/live627/SMF2.1/commit/1c85fd040d5873bdd5204eec31b403debff7517d that should addressees this. Can you say which special characters don't work?

**Comment by m4z** [comment]
After reading the docs on PHP single quoted strings and a bit of testing, I think the culprit breaking the database login was `\`.

To expand on my second problem: If I use a password containing a backslash, the password is entered into `Settings.php` correctly:
```
$db_passwd = '*<)`#?@`<?[\`"+>[?&,<\<{$(*@.,>=~%,!()!"^^~&/^-:*}>*,>?"=#)_\=&+]^~<`^=&$+\"@{#:)/*/!;<[@{~*$\__("/=%?),"@&){*#/&)"=~<>\!*"{`;\:';
```

But the installer can't connect to the DB with that exact password:
```
Critical Error!
Cannot connect to the database server with the supplied data.

If you are not sure about what to type in, please contact your host.
```

If I then try to re-enter the password (or edit the password in the DB and installer WebUI), part of the password is doubled in the config:
```
$db_passwd = '*<)`#?@`<?[\`"+>[?&,<\<{$(*@.,>=~%,!()!"^^~&/^-:*}>*,>?"=#)_\=&+]^~<`^=&$+\"@{#:)/*/!;<[@{~*$\__("/=%?),"@&){*#/&)"=~<>\!*"{`;\:';		#?@`<?[\`"+>[?&,<\<{$(*@.,>=~%,!()!"^^~&/^-:*}>*,>?"=#)_\=&+]^~<`^=&$+\"@{#:)/*/!;<[@{~*$\__("/=%?),"@&){*#/&)"=~<>\!*"{`;\:';
```

I then had to manually remove the password from the config and start the installer from scratch.


With my limited PHP knowledge, I _think_ your patch does exactly the right thing.
Thanks for the quick response and fix, and thanks for this wonderful piece of software!

**Comment by albertlast** [comment]
Single quote is most liky the issue here, since it needs to be escaped,
with another single quote.

**Comment by m4z** [comment]
[The PHP docs say](https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.single) escaping `'` as `\'` works, and the file looks correct with the fix (also, none of my passwords from #5608 today included single quotes IIRC).

**Comment by albertlast** [comment]
I talk from database side,
not php side.
